src/util.rs

```diff
@@ -77,6 +350,61 @@ pub(crate) const fn _round_down_to_next_multiple_of_alignment(
     n & mask
 }
 
+/// Since we support multiple versions of Rust, there are often features which
+/// have been stabilized in the most recent stable release which do not yet
+/// exist (stably) on our MSRV. This module provides polyfills for those
+/// features so that we can write more "modern" code, and just remove the
+/// polyfill once our MSRV supports the corresponding feature. Without this,
+/// we'd have to write worse/more verbose code and leave TODO comments sprinkled
+/// throughout the codebase to update to the new pattern once it's stabilized.
+///
+/// Each trait is imported as `_` at the crate root; each polyfill should "just
+/// work" at usage sites.
+pub(crate) mod polyfills {
+    use core::ptr::{self, NonNull};
+
+    // A polyfill for `NonNull::slice_from_raw_parts` that we can use before our
+    // MSRV is 1.70, when that function was stabilized.
+    //
+    // TODO(#67): Once our MSRV is 1.70, remove this.
+    pub(crate) trait NonNullExt<T> {
+        fn slice_from_raw_parts(data: Self, len: usize) -> NonNull<[T]>;
+    }
+
+    impl<T> NonNullExt<T> for NonNull<T> {
+        #[inline(always)]
+        fn slice_from_raw_parts(data: Self, len: usize) -> NonNull<[T]> {
+            let ptr = ptr::slice_from_raw_parts_mut(data.as_ptr(), len);
+            // SAFETY: `ptr` is converted from `data`, which is non-null.
+            unsafe { NonNull::new_unchecked(ptr) }
+        }
+    }
+
+    // A polyfill for `NonNull::len` that we can use before our MSRV is 1.63,
+    // when that function was stabilized.
+    //
+    // TODO(#67): Once our MSRV is 1.63, remove this.
+    pub(crate) trait NonNullSliceExt<T> {
+        fn len(&self) -> usize;
+    }
+
+    impl<T> NonNullSliceExt<T> for NonNull<[T]> {
+        #[inline(always)]
+        fn len(&self) -> usize {
+            #[allow(clippy::as_conversions)]
+            let slc = self.as_ptr() as *const [()];
```

Comment:
This is subtle. Could you cite the reference that fat pointer conversions preserve length?

Reply:
So fun discovery; there is no such guarantee as of today. I've put up a PR to add one: https://github.com/rust-lang/reference/pull/1417. We'll need to block this PR until that lands.

Reply:
Another option would be to split this PR so that all of the bits that transitively rely on this method are in a follow-up PR. That would include the `try_from` method, so it'd be a good chunk of this PR that we'd need to save for a follow-up.

Reply:
Discussed offline. I've split this and transitive dependencies into https://github.com/google/zerocopy/pull/498.